File: packages/mcp-server/src/tools/meetings.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { pmApi, PmApiError } from '../pm-api.js';

function err(msg: string) { return { content: [{ type: 'text' as const, text: `Error: ${msg}` }] }; }
function ok(data: unknown) { return { content: [{ type: 'text' as const, text: JSON.stringify(data, null, 2) }] }; }

export function registerMeetingTools(server: McpServer) {
  server.tool('list_meetings', 'List meetings for project', {
    meeting_type: z.enum(['standup', 'retro', 'general', 'review']).optional(),
    date_from: z.string().optional().describe('YYYY-MM-DD'),
    date_to: z.string().optional().describe('YYYY-MM-DD'),
    limit: z.number().optional(),
  }, async ({ meeting_type, date_from, date_to, limit }) => {
    try {
      const params = new URLSearchParams();
      if (meeting_type) params.set('meeting_type', meeting_type);
      if (date_from) params.set('date_from', date_from);
      if (date_to) params.set('date_to', date_to);
      if (limit !== undefined) params.set('limit', String(limit));
      const query = params.toString();
      const data = await pmApi(`/api/meetings${query ? `?${query}` : ''}`);
      return ok(data);
    } catch (e) {
      return err(e instanceof PmApiError ? e.message : String(e));
    }
  });

  server.tool('get_meeting', 'Get meeting details', {
    meeting_id: z.string(),
  }, async ({ meeting_id }) => {
    try {
      const data = await pmApi(`/api/meetings/${meeting_id}`);
      return ok(data);
    } catch (e) {
      return err(e instanceof PmApiError ? e.message : String(e));
    }
  });

  server.tool('create_meeting', 'Create meeting', {
    title: z.string(),
    meeting_type: z.enum(['standup', 'retro', 'general', 'review']).optional(),
    date: z.string().optional().describe('ISO datetime'),
    duration_min: z.number().optional(),
    participants: z.array(z.object({ name: z.string() })).optional(),
    created_by: z.string().optional().describe('team_member ID'),
  }, async ({ title, meeting_type, date, duration_min, participants, created_by }) => {
    try {
      const body: Record<string, unknown> = { title };
      if (meeting_type) body.meeting_type = meeting_type;
      if (date) body.date = date;
      if (duration_min !== undefined) body.duration_min = duration_min;
      if (participants) body.participants = participants;
      if (created_by) body.created_by = created_by;
      const data = await pmApi('/api/meetings', { method: 'POST', body: JSON.stringify(body) });
      return ok(data);
    } catch (e) {
      return err(e instanceof PmApiError ? e.message : String(e));
    }
  });

  server.tool('update_meeting', 'Update meeting', {
    meeting_id: z.string(),
    title: z.string().optional(),
    meeting_type: z.enum(['standup', 'retro', 'general', 'review']).optional(),
    date: z.string().optional(),
    duration_min: z.number().optional(),
    participants: z.array(z.object({ name: z.string() })).optional(),
    raw_transcript: z.string().optional(),
    ai_summary: z.string().optional(),
    decisions: z.array(z.object({ id: z.string(), text: z.string(), owner: z.string().optional() })).optional(),
    action_items: z.array(z.object({ id: z.string(), text: z.string(), assignee: z.string().optional(), due_date: z.string().optional(), status: z.string().optional() })).optional(),
  }, async ({ meeting_id, ...updates }) => {
    try {
      const body: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(updates)) {
        if (v !== undefined) body[k] = v;
      }
      const data = await pmApi(`/api/meetings/${meeting_id}`, { method: 'PUT', body: JSON.stringify(body) });
      return ok(data);
    } catch (e) {
      return err(e instanceof PmApiError ? e.message : String(e));
    }
  });

  server.tool('delete_meeting', 'Soft delete meeting', {
    meeting_id: z.string(),
  }, async ({ meeting_id }) => {
    try {
      const data = await pmApi(`/api/meetings/${meeting_id}`, { method: 'DELETE' });
      return ok(data);
    } catch (e) {
      return err(e instanceof PmApiError ? e.message : String(e));
    }
  });

  server.tool('trigger_ai_summary', 'Trigger AI structuring on meeting transcript', {
    meeting_id: z.string(),
  }, async ({ meeting_id }) => {
    try {
      const data = await pmApi(`/api/meetings/${meeting_id}/summary`, {
        method: 'POST',
        body: JSON.stringify({}),
      });
      return ok(data);
    } catch (e) {
      return err(e instanceof PmApiError ? e.message : String(e));
    }
  });
}

File: backend/sprintable_mcp/tools/meetings.py
"""미팅 관련 MCP 도구 (6개)."""
from __future__ import annotations

from typing import Literal

from mcp.types import TextContent

from ..api_client import client
from ..response import err, ok
from ..schemas import SprintableInput

MeetingType = Literal["standup", "retro", "general", "review"]


class ListMeetingsInput(SprintableInput):
    meeting_type: MeetingType | None = None
    date_from: str | None = None
    date_to: str | None = None
    limit: int | None = None


class MeetingIdInput(SprintableInput):
    meeting_id: str


class CreateMeetingInput(SprintableInput):
    title: str
    meeting_type: MeetingType | None = None
    date: str | None = None
    duration_min: int | None = None
    participants: list[dict] | None = None
    created_by: str | None = None


class UpdateMeetingInput(SprintableInput):
    meeting_id: str
    title: str | None = None
    meeting_type: MeetingType | None = None
    date: str | None = None
    duration_min: int | None = None
    participants: list[dict] | None = None
    raw_transcript: str | None = None
    ai_summary: str | None = None
    decisions: list[dict] | None = None
    action_items: list[dict] | None = None


async def list_meetings(args: ListMeetingsInput) -> list[TextContent]:
    """프로젝트 미팅 목록 조회."""
    params: dict = {"project_id": client.project_id}
    if args.meeting_type:
        params["meeting_type"] = args.meeting_type
    if args.date_from:
        params["date_from"] = args.date_from
    if args.date_to:
        params["date_to"] = args.date_to
    if args.limit is not None:
        params["limit"] = str(args.limit)
    try:
        return ok(await client.get("/api/v2/meetings", params=params))
    except Exception as exc:
        return err(str(exc))


async def get_meeting(args: MeetingIdInput) -> list[TextContent]:
    """미팅 상세 조회."""
    try:
        return ok(await client.get(f"/api/v2/meetings/{args.meeting_id}"))
    except Exception as exc:
        return err(str(exc))


async def create_meeting(args: CreateMeetingInput) -> list[TextContent]:
    """미팅 생성."""
    body: dict = {"title": args.title, "project_id": client.project_id}
    for field in ("meeting_type", "date", "duration_min", "participants", "created_by"):
        val = getattr(args, field)
        if val is not None:
            body[field] = val
    try:
        return ok(await client.post("/api/v2/meetings", json=body))
    except Exception as exc:
        return err(str(exc))


async def update_meeting(args: UpdateMeetingInput) -> list[TextContent]:
    """미팅 수정."""
    updates: dict = {}
    for field in ("title", "meeting_type", "date", "duration_min", "participants",
                  "raw_transcript", "ai_summary", "decisions", "action_items"):
        val = getattr(args, field)
        if val is not None:
            updates[field] = val
    try:
        return ok(await client.put(f"/api/v2/meetings/{args.meeting_id}", json=updates))
    except Exception as exc:
        return err(str(exc))


async def delete_meeting(args: MeetingIdInput) -> list[TextContent]:
    """미팅 소프트 삭제."""
    try:
        return ok(await client.delete(f"/api/v2/meetings/{args.meeting_id}"))
    except Exception as exc:
        return err(str(exc))


async def trigger_ai_summary(args: MeetingIdInput) -> list[TextContent]:
    """미팅 AI 요약 생성 트리거."""
    try:
        return ok(await client.post(f"/api/v2/meetings/{args.meeting_id}/summary", json={}))
    except Exception as exc:
        return err(str(exc))
